x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go

```diff
@@ -191,14 +206,50 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
 
 		for namespace, namespaceDetails := range namespaceDetailTotal {
 			m.logger.Debugf("Collected metrics from namespace %s", namespace)
-
 			// filter listMetricsOutput by detailed configuration per each namespace
 			filteredMetricWithStatsTotal := filterListMetricsOutput(listMetricsOutput, namespace, namespaceDetails)
 
 			// get resource type filters and tags filters for each namespace
 			resourceTypeTagFilters := constructTagsFilters(namespaceDetails)
 
-			eventsWithIdentifier, err := m.createEvents(svcCloudwatch, svcResourceAPI, filteredMetricWithStatsTotal, resourceTypeTagFilters, regionName, startTime, endTime)
+			//Check whether namespace is APIGW
+			useonlyrest := false
+			if len(resourceTypeTagFilters) == 1 {
+				for key := range resourceTypeTagFilters {
+					if strings.Compare(strings.ToLower(key), checkresource_type_lower) == 0 {
+						useonlyrest = true
+					}
+				}
+			}
+			if strings.Contains(strings.ToLower(namespace), checkns_lower) {
+				// inforestapi includes only Rest APIs
+				if useonlyrest {
+					infoapi, err = aws.GetAPIGatewayRestAPIOutput(APIClients.Apigateway, config.LimitRestAPI)
+					if err != nil {
+						m.Logger().Errorf("could not get rest apis output: %v", err)
+					}
+				} else {
+					// infoapi  includes only Rest APIs
+					// apiGatewayAPI includes only WebSocket and HTTP APIs
+					infoapi, err = aws.GetAPIGatewayRestAPIOutput(APIClients.Apigateway, config.LimitRestAPI)
+					if err != nil {
+						m.Logger().Errorf("could not get rest apis output: %v", err)
+					}
+
+					apiGatewayAPI, err := aws.GetAPIGatewayAPIOutput(APIClients.Apigatewayv2)
+					if err != nil {
+						m.Logger().Errorf("could not get http and websocket apis output: %v", err)
+					}
+					if len(apiGatewayAPI) > 0 {
+						maps.Copy(infoapi, apiGatewayAPI)
+					}
+
+				}
+
```

Comment:
In general, I try not to have specific AWS service API calls in the "generic" Fetch() function.

The Fetch() function handles metric collection for any metric namespace. I would try not to have conditional behavior based on a specific namespace ("AWS/ApiGateway" in this case) in the main code path.

Reply:
This is what `if strings.Contains(strings.ToLower(namespace), checkns_lower)`  does checks specific namespace == "AWS/ApiGateway"